### DISCUSSION TITLE
How to simplify and improve my form?

### DISCUSSION THREAD

**Discussion by ghost** [alpinejs/alpine#1348] [1. Help] [answered]
Hi,

first of all, I am new to frontend development and try building an API for a Network Automation project.

For the frontend I made this rather complicated form with Tailwind and Alpine: https://codepen.io/freilaeufer/pen/QWpqGZB

The data for each field should later be fetched from an API (hard coded at the moment for test purpose). 

It is possible to add a new range definition by clicking the add button on top or remove a definition by clicking the remove button (only visible after a definition was added). After clicking submit, the form is send to an API. 

The final format should be like this:

[ {from: "xxx", to: "xxx", type: "xxx", vlan: "xxx", voice: "xxx"},  {from: "xxx", to: "xxx", type: "xxx", vlan: "xxx", voice: "xxx"}, ... ]

So my questions are how can I improve my code, to:

- simplify the whole code base (it is very messy and has a lot of repetition)
- dynamically set the x-model attribute for each iteration to get the format I mentioned above

Or is there a more simple way of doing this kind of form? 

Thank you!

**Comment by KevinBatdorf** [comment, marked as answer]
It just depends on how complex you're going to get. if this is it then your code is fine. If you have more complexity you could extract the data out and define the input types, something like:

```json
{
  "id": "some-id",
  "name": "Some name",
  "label": "Select an option",
  "type": "select",
  "options": [
    "option-1"
    "option-2"
  ]
}
```
Then Loop over the data and have html templates to stub them out
```html
<div x-data="{ myForm: [] }">
<template x-for="data in myData">
  <div>
    <template x-if="data.type === 'select">
      <select :name="data.name" x-model="myForm[name]">
        <!-- etc -->
```
I think this will work. I've done something similar in the past. You may need to tweak the `x-model` part (just reply with a demo and I can help tweak it if you want).

> Ah and also how can I set the col-span attribute dynamic using :class="" ? Let`s say only the first two selections get cols-span-4?

You can add a conditional, like `:class="isThisTrue ? 'class-one' : 'class-two'"` or `:class="{'class-one': isThisTrue()}"`

**Reply by ghost** [reply]
Hi Kevin, 

thanks for your reply.

I could adjust my example with your suggestions, however I couldn't implement both functionalities (select options based on a list and adding or removing fields) together.

So I created two new separated examples:

1. Select options: https://codepen.io/freilaeufer/pen/qBrPvNa
2. Adding or removing fields: https://codepen.io/freilaeufer/pen/ZEeXPpg

> I think this will work. I've done something similar in the past.

Any chance you could share that as an example?

> You may need to tweak the x-model part (just reply with a demo and I can help tweak it if you want).

I think I will need someone to assist me in order to merge those two examples. Especially how I can set the x-model to the corresponding entry within the response object. Currently the response it is just an array, but it should be an array of objects for each row added to the form.

The reponse for the first example code should look like this:

`{ foo: val1, bar: val2, buzz: val3 }`

The reponse for the second example code should look like this:

`[{ foo: val1, bar: val2, buzz: val3 }, { foo: val1, bar: val2, buzz: val3 }]`

EDIT:

Okay, so I could achieve the correct response by changing the code for example 1:

On line 13 in HTML:
`x-model="myForm['${data.name}']"`

On line 26 in JavaScript:
`myForm: { foo: "", bar: "", buz: "" }`

Is this a good way to access the correct value in the object or not?

**Reply by KevinBatdorf** [reply]
You probably want to use `@change` instead of `x-model` so you have more control over the how the data is set. `x-model` is essentially the same for `:value=""` combined with an event like `@input` or `@change` so you can use those instead for more control.

**Reply by ghost** [reply]
Sorry, but I couldn't find much info about `@change` and I'm not exactly sure what you mean. Like I said I am still a beginner.

However I have a working solution which you can find here: https://codepen.io/freilaeufer/pen/abJVVBO

Now I have the following problems:

1. The x-model will be overwritten every time a new row is added, so it looks messy
2. When deleting a row, the index is always the number of the last row, so for example I can't delete a row between

Beside that it looks like my code is working or at least the result is correct.

**EDIT: I think I could fix all my issues and now everything is working perfectly!**

Anyway Kevin, would you mind checking my code and maybe give me some hints? Thanks.
